docs: add compactor-scheduler docs (#16224)
grafana/mimir: docs/sources/mimir/references/architecture/components/compactor-scheduler/index.md

## Benefits of using the compactor-scheduler

Compared to standalone mode, scheduler mode:

- Spreads work evenly across the whole compactor fleet. Because jobs aren't tied to the compactors that own a tenant in the hash ring, no compactor sits idle while others are backlogged, which improves utilization and, with properly sized compactors, reduces the overall time to compact.
- Ensures tenant fairness: pending jobs are distributed with a round-robin across tenants, so a tenant with many jobs can't starve the others.
- Retries failed jobs more efficiently. In standalone mode, when a job fails, the compactor restarts the compaction of the whole tenant, and reaches the failed job again only after re-planning and re-walking the preceding jobs. In scheduler mode, a failed job returns to the queue, and the next available compactor leases just that job.

> - Allows scaling compactors based on the amount of pending work, because the compactor-scheduler knows the queue of outstanding jobs.

Reviewer: ```suggestion
- Distributes work dynamically through queuing rather than through a hash ring. If a job is available than any available compactor can lease it.
- Deduplicates compaction planning across compactor replicas. Planned compaction jobs are instead shared through the compactor-scheduler. This reduces the number of object storage operations required for compaction planning.
- Separates compaction planning from execution. Compaction can progress even if compaction planning degrades.
- Caches block metadata in the metadata cache (link?). Standalone mode cached block metadata on disk, but this meant scaling up caused block metadata to be refetched.
- Autoscale based on compaction job information like total bytes rather than second-order signals like CPU usage.
- Retries failed jobs without requiring replanning.
```
Wording isn't precise, but I tried to highlight distinct benefits. I think avoiding talking about tenant fairness is best because shuffle sharding does that too, but requires tuning tenant shard sizes.

Author: Applied your suggestion mostly as is, with a few edits. Also added a short section about caching a few lines below.

### Job leases

Jobs are leased to compactors, not permanently assigned. If a compactor doesn't report progress on a job for longer than `-compactor-scheduler.lease-duration`, the compactor-scheduler makes the job available for other compactors to lease. A job that has been leased more than `-compactor-scheduler.repeated-failure-report-threshold` times without completing is reported as a repeated failure, and once it has been leased `-compactor-scheduler.max-leases` times it is removed from the queue. Planning jobs are exempt from these limits and are always retried. A discarded job is still re-planned and re-enqueued on the next planning interval, unless it no longer exists (for example, if the tenant has been deleted or the blocks were marked for deletion).

### Block metadata cache

Both planning jobs and compaction jobs read block metadata from object storage. Compactors can cache this metadata in a shared external cache, configured with the `-compactor.scheduler-client.metadata-cache.*` parameters, to reduce object storage operations.

## State

> The compactor-scheduler persists its job queues to local disk, in bbolt databases stored under `-compactor-scheduler.bbolt.dir`. The state is sharded across multiple database files, configured by `-compactor-scheduler.bbolt.shard-count`. Changing the shard count triggers an automatic migration of the state at the next startup.

Reviewer: ```suggestion
The compactor-scheduler persists job information to local disk, in bbolt databases stored under `-compactor-scheduler.bbolt.dir`. The state is sharded across multiple database files, configured by `-compactor-scheduler.bbolt.shard-count`. Changing the shard count triggers an automatic migration of the state at the next startup.
```

> After a restart, the compactor-scheduler recovers the queues from disk. When it starts with no recovered state, it delays planning for a few maintenance intervals, defined by `-compactor-scheduler.maintenance-intervals-before-cold-start-planning`, to avoid enqueueing jobs that duplicate work still in progress. After every restart, it also waits a few maintenance intervals, defined by `-compactor-scheduler.maintenance-intervals-before-lease-expiration`, before enforcing lease expiration: lease renewals aren't persisted, so this prevents re-leasing jobs that compactors are still working on. If the persisted state becomes corrupted, it can be wiped and rebuilt: refer to the [recovery steps](../../../../manage/mimir-runbooks/#MimirCompactorSchedulerUnreachable) in the runbooks.

Reviewer: ```suggestion
After a restart, the compactor-scheduler recovers state from disk. When it starts with no recovered state, it delays planning for a few maintenance intervals, defined by `-compactor-scheduler.maintenance-intervals-before-cold-start-planning`, to avoid enqueuing jobs that duplicate work still in progress. After every restart, it also waits a few maintenance intervals, defined by `-compactor-scheduler.maintenance-intervals-before-lease-expiration`, before enforcing lease expiration to provide a grace period for compactors. If the persisted state becomes corrupted, it can be wiped and rebuilt: refer to the [recovery steps](../../../../manage/mimir-runbooks/#MimirCompactorSchedulerUnreachable) in the runbooks.
```
